iotdb/src/main/java/org/apache/iotdb/db/sync/sender/FileSenderImpl.java

```diff
@@ -0,0 +1,507 @@
+        file.createNewFile();
+        uuid = generateUUID();
+        out.write(uuid.getBytes());
+      } catch (IOException e) {
+        LOGGER.error("Cannot write UUID to file {}", file.getPath());
+        throw new IOException(e);
+      }
+    } else {
+      try (BufferedReader bf = new BufferedReader((new FileReader(uuidPath)))) {
+        uuid = bf.readLine();
+      } catch (IOException e) {
+        LOGGER.error("Cannot read UUID from file{}", file.getPath());
+        throw new IOException(e);
+      }
+    }
+    boolean legalConnection;
+    try {
+      legalConnection = serviceClient.checkIdentity(uuid,
+          InetAddress.getLocalHost().getHostAddress());
+    } catch (Exception e) {
+      LOGGER.error("Cannot confirm identity with receiver");
+      throw new SyncConnectionException(e);
+    }
+    return legalConnection;
+  }
+
+  private String generateUUID() {
+    return Constans.SYNC_CLIENT + UUID.randomUUID().toString().replaceAll("-", "");
+  }
+
+  /**
+   * Create snapshots for valid files.
+   */
+  @Override
+  public Set<String> makeFileSnapshot(Set<String> validFiles) throws IOException {
+    Set<String> validFilesSnapshot = new HashSet<>();
+    try {
+      for (String filePath : validFiles) {
+        String snapshotFilePath = SyncUtils.getSnapshotFilePath(filePath);
+        validFilesSnapshot.add(snapshotFilePath);
+        File newFile = new File(snapshotFilePath);
+        if (!newFile.getParentFile().exists()) {
+          newFile.getParentFile().mkdirs();
+        }
+        Path link = FileSystems.getDefault().getPath(snapshotFilePath);
+        Path target = FileSystems.getDefault().getPath(filePath);
+        Files.createLink(link, target);
+      }
+    } catch (IOException e) {
+      LOGGER.error("Can not make fileSnapshot");
+      throw new IOException(e);
+    }
+    return validFilesSnapshot;
+  }
+
+  /**
+   * Transfer data of a storage group to receiver.
+   *
+   * @param fileSnapshotList list of sending snapshot files in a storage group.
+   */
+  public void syncData(Set<String> fileSnapshotList) throws SyncConnectionException {
+    try {
+      int successNum = 0;
+      for (String snapshotFilePath : fileSnapshotList) {
+        successNum++;
+        File file = new File(snapshotFilePath);
+        List<String> filePathSplit = new ArrayList<>();
+        String os = System.getProperty("os.name");
+        if (os.toLowerCase().startsWith("windows")) {
+          String[] name = snapshotFilePath.split(File.separator + File.separator);
+          filePathSplit.add(name[name.length - 2]);
+          filePathSplit.add(name[name.length - 1]);
+        } else {
+          String[] name = snapshotFilePath.split(File.separator);
+          filePathSplit.add(name[name.length - 2]);
+          filePathSplit.add(name[name.length - 1]);
+        }
+        while (true) {
+          // Sync all data to receiver
+          byte[] buffer = new byte[Constans.DATA_CHUNK_SIZE];
+          int data;
```

Comment:
The naming is confusing.

Reply:
Do you mean DATA_CHUNK_SIZE?

Reply:
I mean `int data`, but it is actually it is the length of the data you have read.

Reply:
yeah, maybe its name is a little bit confusing.